network_simulator/link.py

```diff
@@ -95,3 +164,118 @@ def __eq__(self, other):
+        """
+        Returns link cost. For now, just returns static propagation delay.
+        :return: float of link delay time in seconds
+        """
+
+        # TODO consider average wait time for last N packets
+        return self.static_delay_sec
+
+    def push(self, packet, global_clock_sec):
+        """
+        Pushes packet onto own LinkBuffer
+
+        :param Packet packet: Packet to push
+        :param global_clock_sec: store when packet pushed
+        :return: True if pushed, False if not
+        """
+        return self.link_buffer.push(packet, global_clock_sec)
+
+
+
+class LinkSendEvent(Event):
+    """
+    Event representing a packet arriving at the Link
+    """
+
+    def __init__(self, link):
+        """
+        :ivar link: link that must send its next packet
+        :ivar buff_elem: temporary variable containing LinkBufferElement to send
+        """
+        self.link = link
+
+    def run(self, main_event_loop, statistics):
+        """
+        runs by popping off top packet
+        :param MainEventLoop main_event_loop: event loop for global timing
+        :param Statistics statistics: the Statistics to update
+        """
+
+        # sanity check
+        if(self.link.queue.empty()):
+            logger.warning("LinkSendEvent while Link empty!")
+            return
+
+        # now pop
+        self.buffer_elem = self.link.pop()
+
+        # update statistics only if not RouterPacket
+        if(not isinstance(self.buffer_elem.packet, RouterPacket)):
+            statistics.link_packet_transmitted(link, buffer_elem.packet, \
+                    main_event_loop.global_cloc_sec)
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedules new Events. This is called immediately after run().
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        # schedule a RouterReceivedPacket with propagation + transmission delay
+        main_event_loop.schedule_event_with_delay(RouterReceivedPacketEvent(0),\
+                self.link.static_delay_sec + self.packet.size_bits / \
+                    self.link.capacity_bps)
+            # TODO(Laksh): RouterReceivedPacketEvent signature
+            # recommended signature RouterReceivedPacketEvent(router, packet)?
+
+        if self.link.link_buffer.queue.empty():
+            # if link_buffer.queue is empty, then link is no longer busy
+            self.link.busy = False
+        else:
+            # queue new event with just transmission delay
+            main_event_loop.schedule_event_with_delay(LinkSendEvent(self.link),\
+                    self.buffer_elem.packet.size_bits / self.link.capacity_bps)
+
+
+
+class DeviceToLinkEvent(Event):
+    """
+    Event representing a packet arriving at the Link
+    """
+
+    def __init__(self, packet, link):
+        """
+        :ivar packet: packet arriving at Link
+        :ivar link: link that packet is arriving at
```

Comment:
Again, you need to know the direction this Packet is going in (is it going to end_1_device or end_2_device). This needs to be added to the LinkBufferElement